Fix Kerberos ticket refresh

hashhar opened the pull request:
The Hadoop UGI class handles ticket refresh only if the Subject is not provided externally. For external Subject UGI expects the refresh will be handled by the creator of the Subject which in our case we did not do.

Because of this before this change any Trino query which ran longer than the ticket_lifetime failed with errors like

    GSS initiate failed [Caused by GSSException: No valid credentials
    provided (Mechanism level: Failed to find any Kerberos tgt)].

In Hadoop code the UGI instance also gets re-used in some places (e.g. DFSClient) which means we cannot just create a new UGI with refreshed credentials and return that since other parts of code will keep using the old UGI with expired credentials. So the fix is to create a new UGI, extract the credentials from it and update the existing UGI's credentials with them so that all users of the existing UGI also observe the new valid credentials.

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Hive
* Fix possible query failure with a Kerberized Hive connector when the query executes longer than the Kerberos ticket lifetime. ({issue}`16680`)
```

kokosing commented:
Is it still draft?

hashhar commented:
draft only because I want to remember to remove the stress test commit

hashhar commented on testing/trino-product-tests-launcher/src/main/resources/docker/presto-product-tests/common/hadoop-kerberos/krb5.conf:
```diff
@@ -0,0 +1,26 @@
+ dns_lookup_kdc = false
+ forwardable = true
+ # low ticket_lifetime to make sure refresh code in Trino is tested
+ ticket_lifetime = 3s
```
what do you think about a higher value like 1m and a dedicated test which runs a query longer than 1m (very long query + cancel query at 2m and verify failure message is what we expect).

Current way of testing assumes we have some query running longer than ticket_lifetime that triggers ticket renewal.

Praveen2112 replied:
Maybe we could have a dedicated test which runs a query longer than the lifetime to ensure that we don't fail due to kerberos error.

hashhar commented:
Added a better test, it fails without the fix and passes only with the fix applied.

hashhar commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveConnectorKerberosSmokeTest.java:
```diff
@@ -0,0 +1,106 @@
+        String sql = "CREATE TABLE orders AS SELECT * FROM tpch.sf1000.orders";
+        Future<?> queryExecution = executor.submit(() -> onTrino().executeQuery(sql));
+
+        // 3x of ticket_lifetime as configured in hadoop-kerberos krb5.conf, sufficient to cause at-least 1 ticket expiry
+        SECONDS.sleep(90L);
+        cancelQuery(sql);
```
I'm trying to change this test to instead:

- Run query on executor
- Cancel query after some-time, ignoring failures
- Somehow get the stack trace/cause of the query failure and assert that it failed only because of cancellation

The problem with current assertion is that in case of kerberos failure query fails before the sleep finishes so `cancelQuery` cannot find the query and throws an `IllegalStateException("Query did not reach running state or finished too quickly")` instead of verifying the failure message of the failed query.

hashhar commented on testing/trino-product-tests-launcher/src/main/resources/docker/presto-product-tests/common/hadoop-kerberos/krb5.conf:
```diff
@@ -0,0 +1,25 @@
+ forwardable = true
+ allow_weak_crypto = true
+ # low ticket_lifetime to make sure refresh code in Trino is tested
+ ticket_lifetime = 30s
```
Note that I had to remove `renew_lifetime` from here to avoid hitting https://github.com/trinodb/trino/issues/11251#issuecomment-1061566447.

We can add it back if we have https://github.com/trinodb/docker-images/pull/162 but I don't see any reason to have it since our code doesn't care whether tickets are renewable or not since it always performs a "relogin" instead of renewal.

hashhar replied:
reminder to self, https://github.com/trinodb/docker-images/pull/162 is moving forward.

hashhar commented:
@electrum @Praveen2112 this is ready for another round of review, I've simplified the test and fixed the failures due to having `renew_lifetime` in the krb5.conf.

Praveen2112 commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/suite/suites/Suite2.java:
```diff
@@ -38,16 +38,16 @@ public List<SuiteTestRun> getTestRuns(EnvironmentConfig config)
                         .build(),
                 testOnEnvironment(EnvSinglenodeKerberosHiveNoImpersonationWithCredentialCache.class)
-                        .withGroups("configured_features", "storage_formats", "hdfs_no_impersonation")
+                        .withGroups("configured_features", "storage_formats", "hdfs_no_impersonation", "hive_kerberos")
```
Do we need to run this test on kerberos with credentialCache environment

hashhar replied:
Probably not since the cache's ticket lifetime is larger than the time the test waits (30m vs 30seconds).

I'll remove.

hashhar replied:
removed from credential cache tests

Praveen2112 commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveConnectorKerberosSmokeTest.java:
```diff
@@ -0,0 +1,97 @@
+        String sql = "CREATE TABLE orders AS SELECT * FROM tpch.sf1000.orders";
+        Future<?> queryExecution = executor.submit(() -> onTrino().executeQuery(sql));
+
+        // 2x of ticket_lifetime as configured in hadoop-kerberos krb5.conf, sufficient to cause at-least 1 ticket expiry
```
Is it possible to run a query which takes more than 2x - so in this case we could just run the query and wait for it to complete ?

hashhar replied:
It's difficult to control the amount of time the query executes, it varies a lot under concurrency.

One option is to add a query which runs for ~3x time and use a timer to assert that the query took at-least 3x time to finish to make sure that the test actually runs long enough to hit a ticket expiry. The other problem is that if the query keeps running longer for some reason it'll either timeout the test or fail it.

Praveen2112 commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveConnectorKerberosSmokeTest.java:
```diff
@@ -0,0 +1,97 @@
+            throws Exception
+    {
+        // Session properties to reduce memory usage and also make the query run longer
+        onTrino().executeQuery("SET SESSION scale_writers = false");
+        onTrino().executeQuery("SET SESSION task_scale_writers_enabled = false");
```
Will this session be reflected across all the usage of `onTrino`

hashhar replied:
When `onTrino` is closed it closes the connection it was using thus resetting session properties.

I'll need to verify whether the session is shared even across test classes.

cc: @findepi if he knows the answer. I also see lots of other tests use SET SESSION directly without resetting or any form of isolation.

Praveen2112 replied:
So do we close the connection once the a test is completed ?

hashhar replied:
Actually yes, anything we get from TestContext (e.g. `testContext().getDependency(QueryExecutor.class, config)`) gets closed after each test. This is done by tempto, see https://github.com/trinodb/tempto/blob/a2f7ef3b914db2aeb3480cd845929b1ab26103ef/tempto-core/src/main/java/io/trino/tempto/internal/context/GuiceTestContext.java#L135

Thanks @kokosing for explaining how this magic works.

hashhar commented:
refresh != renewal. I kept refresh because existing code uses `getNextRefreshTime` for example.

I do agree however that relogin makes more sense. In a follow-up I can rename the methods as well to use `relogin` to clarify what the code actually does. Would that be fine?

Praveen2112 commented:
Yeah

kokosing commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveConnectorKerberosSmokeTest.java:
```diff
@@ -0,0 +1,97 @@
+
+        // 2x of ticket_lifetime as configured in hadoop-kerberos krb5.conf, sufficient to cause at-least 1 ticket expiry
+        SECONDS.sleep(60L);
+        cancelQueryIfRunning(sql);
```
do we need to cancel the query? Why not simply wait?

hashhar replied:
It runs very very long. See some reasons to not wait in https://github.com/trinodb/trino/pull/16680#discussion_r1148815727
